src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BlobContainerClientProxy.java

```diff
@@ -69,6 +78,91 @@ private BlobContainerClient get(TargetStorageAccount targetStorageAccount, Strin
+            if ((targetStorageAccount == TargetStorageAccount.CFT
+                || targetStorageAccount == TargetStorageAccount.PCQ)
+                && HttpStatus.valueOf(ex.getResponse().getStatusCode()).is4xxClientError()) {
+                sasTokenCache.removeFromCache(destinationContainer);
+            }
+            throw ex;
+        }
+    }
+
+    private void streamContent(
+        BlockBlobClient sourceBlob,
+        String destinationContainer,
+        TargetStorageAccount targetStorageAccount
+    ) {
+        var blobName = sourceBlob.getBlobName();
+        logger.info("Start streaming from blob {} to Container: {}", sourceBlob.getBlobUrl(), destinationContainer);
+        long startTime = System.nanoTime();
+        try (var zipStream = new ZipInputStream(sourceBlob.openInputStream());) {
+            ZipEntry entry;
+
+            while ((entry = zipStream.getNextEntry()) != null) {
+                if (Objects.equals(entry.getName(), ENVELOPE)) {
+                    final BlockBlobClient blockBlobClient =
+                        get(targetStorageAccount, destinationContainer)
+                            .getBlobClient(blobName)
+                            .getBlockBlobClient();
+
+                    ParallelTransferOptions parallelTransferOptions =
+                        new ParallelTransferOptions()
+                            .setBlockSizeLong(BLOCK_SIZE)
+                            .setMaxConcurrency(8)
+                            .setMaxSingleUploadSizeLong(BLOCK_SIZE);
+
+                    try (var blobOutputStream =
+                        blockBlobClient.getBlobOutputStream(
+                            parallelTransferOptions, null, null, null, null)
+                    ) {
+
+                        byte[] envelopeData = new byte[BUFFER_SIZE];
+                        while (zipStream.available() != 0) {
+                            int numBytesRead = zipStream.readNBytes(envelopeData, 0, BUFFER_SIZE);
+                            blobOutputStream.write(envelopeData, 0, numBytesRead);
+                        }
+                        blobOutputStream.flush();
+                    } catch (Exception ex) {
+                        logger.error("Uploading got error. Stream from {} to {}",
+                            sourceBlob.getBlobUrl(),
+                            destinationContainer,
+                            ex
+                        );
+                        throw ex;
+                    }
+                    logger.info(
+                        "Uploading finished for  blob {} to Container: {}, Upload Duration: {} sec",
+                        sourceBlob.getBlobUrl(),
+                        destinationContainer,
+                        TimeUnit.NANOSECONDS.toSeconds(System.nanoTime() - startTime)
+                    );
+
+                    return;
+                }
+            }
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
```

Comment:
Makes sense to use a customized exception.

Reply:
add new exception type